gapis/gfxapi/gles/gles.api

```diff
@@ -515,34 +515,47 @@ sub void ApplyStaticContextState(ref!Context ctx, ref!StaticContextState staticS
 
 sub void ApplyDynamicContextState(ref!Context ctx, ref!DynamicContextState dynamicState) {
   if (dynamicState != null) {
-    backbuffer := ctx.Objects.Framebuffers[0]
-    color_id := backbuffer.ColorAttachments[0].Renderbuffer.ID
-    depth_id := backbuffer.DepthAttachment.Renderbuffer.ID
-    stencil_id := backbuffer.StencilAttachment.Renderbuffer.ID
-
-    color_buffer := ctx.SharedObjects.Renderbuffers[color_id]
-    color_buffer.Width = dynamicState.BackbufferWidth
-    color_buffer.Height = dynamicState.BackbufferHeight
-    color_buffer.InternalFormat = dynamicState.BackbufferColorFmt
-    ctx.SharedObjects.Renderbuffers[color_id] = color_buffer
-
-    depth_buffer := ctx.SharedObjects.Renderbuffers[depth_id]
-    depth_buffer.Width = dynamicState.BackbufferWidth
-    depth_buffer.Height = dynamicState.BackbufferHeight
-    depth_buffer.InternalFormat = dynamicState.BackbufferDepthFmt
-    ctx.SharedObjects.Renderbuffers[depth_id] = depth_buffer
-
-    stencil_buffer := ctx.SharedObjects.Renderbuffers[stencil_id]
-    stencil_buffer.Width = dynamicState.BackbufferWidth
-    stencil_buffer.Height = dynamicState.BackbufferHeight
-    stencil_buffer.InternalFormat = dynamicState.BackbufferStencilFmt
-    ctx.SharedObjects.Renderbuffers[stencil_id] = stencil_buffer
-
-    if dynamicState.ResetViewportScissor {
-      ctx.FragmentOperations.Scissor.Box.Width = dynamicState.BackbufferWidth
-      ctx.FragmentOperations.Scissor.Box.Height = dynamicState.BackbufferHeight
-      ctx.Rasterization.Viewport.Width = dynamicState.BackbufferWidth
-      ctx.Rasterization.Viewport.Height = dynamicState.BackbufferHeight
+    if 0 in ctx.Objects.Framebuffers {
```

Comment:
> am not happy about this - I do want to write the whole API file based on the assumption the data might get inconsistent at any point.

The code generated from the API cannot crash, full-stop.

I understand that you don't want to add seat-belts everywhere through the API files, but NPEs are not acceptable for _any_ sequence of calls (typical or unusual). Longer term we should be fuzzing the API to prevent panics from occurring, ever.

For this particular case, it would be usual for the framebuffer attachments to be missing, but it is possible once we implement #113, as you just have to delete the first `eglMakeCurrent` which holds the `StaticContextState`. I'm not saying this is the only solution to the problem, but it solves the crashes I encountered (I also wrote a simple stress test of 1000's of concurrent query-wait-cancels to ensure there were no other hidden panics). 

Most of the work I have done with the static analysis was exactly to address this - if we can accurately determine what all the possible entry states are for any command, then we can avoid adding unnecessary seat-belts everywhere.

I've considered attempting to automatically handle things like NPEs, turning them into errors, but each time I've considered it, I've come to the conclusion it's a bad idea. Either the NPE is a failure on our part (which we should fix), or is an error case that we should be reporting with a sensible human readable message.

>  If you kick of a job, the job should finish.

What I failed to add in this PR description are the performance improvements these changes give. Quickly stepping though the commands stacked up walks of state mutation, which could completely stall getting the most recent results. Actually cancelling atom-walks gives noticeable improvements to the responsiveness in the UI.

For that reason, I disagree.

> fine for very expensive jobs to have high level check if the context has been cancelled 

I don't really want to start this argument again, but I'll repeat what I said before - the things that are checking cancellation state are "expensive jobs". The ones that caused these issues were all to do with resolvables - which by nature **are** expensive (otherwise they wouldn't be resolvables in the first place) - and they just happen to be called from `atom.Mutate()`. The obvious problem is that mutate logic depends on the results of these possibly cancelled jobs, and so we have to have some basic handling of context creation in the mutator.

I get that this solution isn't perfect - the cancellation will only be noticed at the end of the `Mutate()` call, and so part of the logic within the `Mutate()` can be using invalid data if the cancellation happens early. The idiomatic way to deal with this is to return and check `errors` everywhere, but the template code generation makes this extremely difficult and I'm worried about runtime / compile time impact. This was the best compromise I could come up with for now. If we finally migrate to the LLVM based tech, then we could probably use exceptions to handle these sorts of cases (including `abort`), which would terminate execution as soon as the cancellation is raised.